reverseproxy/proxy.go

```diff
@@ -44,130 +51,119 @@ func NewReversePaymentProxy(proxyAddress string, nitroEndpoint string, destinati
 	if err != nil {
 		panic(err)
 	}
-	// Creates a reverse proxy that will handle forwarding requests to the destination server
-
-	proxy := &httputil.ReverseProxy{
-		Rewrite: func(r *httputil.ProxyRequest) {
-			// SetURL updates the URL and will update the host header with the proxy
-			// This avoids problems with servers that check the host header against the requestor
-			r.SetURL(destinationUrl)
-		},
-	}
 
-	return &ReversePaymentProxy{
-		server:                server,
-		logger:                logger,
-		nitroClient:           nitroClient,
-		reverseProxy:          proxy,
-		expectedPaymentAmount: expectedPaymentAmount,
+	p := &ReversePaymentProxy{
+		server:         server,
+		logger:         logger,
+		nitroClient:    nitroClient,
+		costPerByte:    costPerByte,
+		destinationUrl: destinationUrl,
+
+		reverseProxy: &httputil.ReverseProxy{},
 	}
-}
 
-// Start starts the proxy server in a goroutine.
-func (p *ReversePaymentProxy) Start() error {
-	// Wire up our proxy to the http handler
-	// This means that p.ServeHTTP will be called for every request
+	// Wire up our handlers to the reverse proxy
+	p.reverseProxy.Rewrite = func(pr *httputil.ProxyRequest) { pr.SetURL(p.destinationUrl) }
+	p.reverseProxy.ModifyResponse = p.handleDestinationResponse
+	p.reverseProxy.ErrorHandler = p.handleError
+	// Wire up our handler to the server
 	p.server.Handler = p
 
-	go func() {
-		p.logger.Info().Msgf("Starting reverse payment proxy listening on %s.", p.server.Addr)
-		p.logger.Info().Msgf("Each request will cost %d wei", p.expectedPaymentAmount.Uint64())
-		if err := p.server.ListenAndServe(); err != http.ErrServerClosed {
-			p.logger.Err(err).Msg("ListenAndServe()")
-		}
-	}()
-
-	return nil
+	return p
 }
 
-// Stop stops the proxy server and closes everything.
-func (p *ReversePaymentProxy) Stop() error {
-	p.logger.Info().Msgf("Stopping reverse payment proxy listening on %s", p.server.Addr)
-	err := p.server.Shutdown(context.Background())
+// ServeHTTP is the main entry point for the reverse payment proxy server.
+// It is responsible for parsing the voucher from the query params and moving it to the request header
+// It then delegates to the reverse proxy to handle rewriting the request and sending it to the destination
+func (p *ReversePaymentProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	v, err := parseVoucher(r.URL.Query(), "")
 	if err != nil {
-		return err
+		p.handleError(w, r, createPaymentError(fmt.Errorf("could not parse voucher: %w", err)))
+		return
 	}
 
-	return p.nitroClient.Close()
+	// We remove the voucher from the query params so it doesn't pollute the URL
+	queryParams := r.URL.Query()
+	removeVoucher(queryParams, "")
+	r.URL.RawQuery = queryParams.Encode()
+
+	// We add the voucher to the request header so we can easily check the cost in the response handler without polluting the URL
+	addVoucher(v, r.Header, HEADER_PREFIX)
```

Comment:
This isn't strictly necessary, we could just keep the voucher on the query params and parse it out from that later. However that means the voucher params would get passed to the destination server, which it may not expect. By moving it to a custom header we still have access to it but it's less likely to cause issues with the destination server.

Reply:
Ooh good idea!